sys/net/application_layer/gcoap/dns.c

```diff
@@ -0,0 +1,683 @@
+        if (res < 0) {
+            DEBUG("gcoap_dns: cannot add credential to GCoAPs DTLS sock\n");
+            return -EBADF;
+        }
+#endif
+        c->tag = creds->tag;
+        c->type = creds->type;
+    }
+    return 0;
+}
+
+void gcoap_dns_cred_remove(credman_tag_t tag, credman_type_t type)
+{
+#if IS_USED(MODULE_GCOAP_DTLS)
+    sock_dtls_t *sock = gcoap_get_sock_dtls();
+    for (unsigned i = 0; i < ARRAY_SIZE(_creds); i++) {
+        if ((_creds[i].tag == tag) && (_creds[i].type == type)) {
+            _remove_cred(sock, &_creds[i]);
+            break;
+        }
+    }
+#else
+    (void)tag;
+    (void)type;
+#endif
+}
+
+void gcoap_dns_server_proxy_reset(void)
+{
+    if (IS_USED(MODULE_GCOAP_DNS_PROXIED)) {
+        mutex_lock(&_client_mutex);
+        _proxy[0] = '\0';
+        mutex_unlock(&_client_mutex);
+    }
+}
+
+int gcoap_dns_server_proxy_set(const char *proxy)
+{
+    int res;
+
+    if (!IS_USED(MODULE_GCOAP_DNS_PROXIED)) {
+        return -ENOTSUP;
+    }
+    res = strlen(proxy);
+
+    if ((unsigned)(res + 1) >= CONFIG_GCOAP_DNS_SERVER_URI_LEN) {
+        return -EINVAL;
+    }
+    mutex_lock(&_client_mutex);
+    strcpy(_proxy, proxy);
+    mutex_unlock(&_client_mutex);
+    return res;
+}
+
+const char *gcoap_dns_server_proxy_get(void)
+{
+    if (_is_proxied()) {
+        return _proxy;
+    }
+    return NULL;
+}
+
+static inline bool _dns_server_uri_isset(void)
+{
+    return _uri[0] != '\0';
+}
+
+#if IS_USED(MODULE_GCOAP_DTLS)
+static void _remove_cred(sock_dtls_t *sock, _cred_t *cred)
+{
+    sock_dtls_remove_credential(sock, cred->tag);
+    credman_delete(cred->tag, cred->type);
+    cred->type = CREDMAN_TYPE_EMPTY;
+    cred->tag = CREDMAN_TAG_EMPTY;
+}
+#endif
+
+static inline bool _is_proxied(void)
+{
+    return IS_USED(MODULE_GCOAP_DNS_PROXIED) && _proxy[0] != '\0';
+}
+
+static int _add_init_block2_opt(coap_pkt_t *pdu)
+{
+    if (CONFIG_GCOAP_DNS_PDU_BUF_SIZE < CONFIG_DNS_MSG_LEN) {
+        coap_block1_t block;
+
+        coap_block_object_init(&block, 0, CONFIG_GCOAP_DNS_BLOCK_SIZE, 0);
+        return coap_opt_add_block2_control(pdu, &block);
+    }
+    return 0;
+}
+
+static int _add_proxy_uri_opt(coap_pkt_t *pdu, const char *proxy_uri)
+{
+    if (_is_proxied()) {
+        return coap_opt_add_proxy_uri(pdu, proxy_uri);
+    }
+    return 0;
+}
+
+static int _add_req_tag_opt(coap_pkt_t *pdu, _req_ctx_t *context)
+{
+    if (CONFIG_GCOAP_DNS_PDU_BUF_SIZE < CONFIG_DNS_MSG_LEN) {
+        return coap_opt_add_opaque(pdu, 292, (uint8_t *)&context->req_tag,
+                                   sizeof(context->req_tag));
+    }
+    return 0;
+}
+
+static int _add_remaining_options(coap_pkt_t *pdu, const char *proxy_uri, _req_ctx_t *context)
+{
+    if (_add_proxy_uri_opt(pdu, proxy_uri) < 0) {
+        DEBUG("gcoap_dns: unable to add Proxy-URI option to request\n");
+        return -ENOBUFS;
+    }
+    /* add request tag to distinguish multiple blockwise requests on-the-air */
```

Comment:
Correction: There *can* only ever be one caller to the internals internals of address resolution, because `_client_mutex` is locked in `gcoap_dns_query`. (I only found that when I wondered why the global `_remote` could be used).

I see three ways to make things better:
* Strictly become single-request-only. Then, Request-Tag can go away .. although we'd need it with DTLS, and under some conditions with OSCORE ... so the comment here should say something like
  > While there is only ever one request in flight, request tags are necessary for OSCORE and DTLS to ensure body integrity.
* Allow multiple concurrent requests (necessitating a move of `_remote` into `_req`, but possibly a lot of different things too, and it's kind of violating CoAP requirements unless we regard different threads as different users.
* Leave things in a mixed state, and add a paragraph somewhere for whoever next touches this.

Reply:
> * Strictly become single-request-only. Then, Request-Tag can go away .. although we'd need it with DTLS, and under some conditions with OSCORE ... so the comment here should say something like
>       > While there is only ever one request in flight, request tags are necessary for OSCORE and DTLS to ensure body integrity.

Can you give an example where we would need it for DTLS / OSCORE? I am inclined to just remove the request tag for now and find a more general mechanism in the future.

Reply:
There are possible request-rearrangement attacks to both DTLS and OSCORE. (In DTLS, they're relatively easy; in OSCORE they are harder but still not ruled out). With those, an attacker can (in our case) swap out the first (few) blocks of a request, and then produce a situation for which client and server disagree on the request body without the cryptographic checks triggering.

A 16-bit Request-Tag is not the sharpest tool to use (the more precise conditions [require some tracking](https://www.rfc-editor.org/rfc/rfc9175.html#name-body-integrity-based-on-pay)), but I think it'd make attacks impractical.

Reply:
OK, then let's make the request tag dependendend on `gcoap_dtls` / `oscore` (in https://github.com/miri64/RIOT/tree/gcoap_dns/enh/oscore).

Reply:
That's a viable option. (I don't see how these mixups would happen accidentally, and if there was an attacker the Request-Tag would do the user no good in unsecured mode).